### PR TITLE
fix: respect graceful mode in Android SDK

### PR DESCRIPTION
**Pull request by typotter**
fixes FF-3565
[Related PR](https://github.com/Eppo-exp/sdk-common-jdk/pull/68)

### CODE REVIEW

**Review by felipecsl — APPROVED** [review]

**Review by aarsilv — APPROVED** [review]

File: eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java [hunk 1]
```diff
@@ -258,6 +259,98 @@ public void testErrorGracefulModeOff() {
...
+    EppoClient.Builder clientBuilder =
+        new EppoClient.Builder(DUMMY_API_KEY, ApplicationProvider.getApplicationContext())
+            .forceReinitialize(true)
+            .isGracefulMode(true);
+
+    // Initialize and no exception should be thrown.
+    try {
+      clientBuilder.buildAndInitAsync().get();
+    } catch (Exception e) {
+      fail("Unexpected exception thrown: " + e.getMessage());
```
Comment (felipecsl) [inline eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java:292]: nit: you don't really need the try-catch, as any exception will fail the test, but not a big deal if you like it

Comment (typotter) [reply eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java]: 👍

File: eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java [hunk 2]
```diff
@@ -258,6 +259,98 @@ public void testErrorGracefulModeOff() {
...
+  public void testClientMakesDefaultAssignmentsAfterFailingToInitializeNonGracefulMode() {
+    // Set up bad HTTP response
+    setBaseClientHttpClientOverrideField(mockHttpError());
+
+    EppoClient.Builder clientBuilder =
+        new EppoClient.Builder(DUMMY_API_KEY, ApplicationProvider.getApplicationContext())
+            .forceReinitialize(true)
+            .isGracefulMode(false);
+
+    // Initialize and no exception should be thrown.
```
[the remark this answers is not in the document]
Comment (typotter) [reply eppo/src/androidTest/java/cloud/eppo/android/EppoClientTest.java]: ooof.

File: eppo/src/main/java/cloud/eppo/android/EppoClient.java [hunk 3]
```diff
@@ -275,16 +276,37 @@ public CompletableFuture<EppoClient> buildAndInitAsync() {
...
-      } catch (ExecutionException | InterruptedException e) {
-        throw new RuntimeException(e);
+      } catch (ExecutionException | InterruptedException | CompletionException e) {
+        // If the exception was an `EppoInitializationException`, we know for sure that
+        // `buildAndInitAsync` logged it (and wrapped it with a RuntimeException) which was then
+        // wrapped by `CompletableFuture` with a `Completion Exception`
+        if (e instanceof CompletionException
+            && (e.getCause() != null && e.getCause() instanceof RuntimeException)
+            && (e.getCause().getCause() != null
+                && e.getCause().getCause() instanceof EppoInitializationException)) {
```
[the remark this answers is not in the document]
Comment (typotter) [reply eppo/src/main/java/cloud/eppo/android/EppoClient.java]: Cleaned up the condition a bit. I was not aware that instanceOf covers `!= null` as well, so much more readable.

**Comment by typotter** [comment]
Thanks Felipe.

**Pull request merged**